gpytorch/priors/lkj_prior.py

```diff
@@ -61,6 +61,47 @@ def log_prob(self, X):
         log_diag_sum = psd_safe_cholesky(X, upper=True).diagonal(dim1=-2, dim2=-1).log().sum(-1)
         return self.C + (self.eta - 1) * 2 * log_diag_sum
 
+    def _rsample(self, sample_shape=None, return_covariance=False):
+        # this sampling method comes from
+        # https://github.com/rmcelreath/rethinking/blob/2acf2fd7b01718cf66a8352c52d001886c7d3c4c/R/distributions.r#L214
+        if sample_shape is None:
+            sample_shape = torch.Size()
+
+        N = self.n.item()
+
+        alpha = self.eta + (N - 1) / 2
+        r12 = 2.0 * Beta(alpha, alpha).rsample(sample_shape) - 1
+        R = torch.zeros(*sample_shape, N, N, device=self.eta.device, dtype=self.eta.dtype)
+
+        R[..., 0, 0] = 1.0
+        R[..., 0, 1] = r12
+        R[..., 1, 1] = (1 - r12.pow(2)).pow(0.5)
+
+        for m in range(1, N):
+            alpha = alpha - 0.5
+            y = Beta(m / 2, alpha).rsample(sample_shape)
+
+            # uniform on a hypersphere
+            z = torch.randn(*sample_shape, m, device=self.eta.device, dtype=self.eta.dtype)
+            z = z / z.pow(2.0).sum(-1, keepdim=True).pow(0.5)
+
+            R[..., :m, m] = y.pow(0.5).unsqueeze(-1) * z
+            R[..., m, m] = (1.0 - y).pow(0.5)
+
+        sample_as_covariance = R.matmul(R.transpose(-1, -2))
+
+        if not return_covariance:
+            # we seem to need to back-convert the resulting matrix into a correlation matrix
+            inverse_diagonal = torch.diag_embed(
+                torch.diagonal(sample_as_covariance, dim1=-2, dim2=-1).clamp(min=1e-6).pow(-0.5)
+            )
+            return inverse_diagonal.matmul(sample_as_covariance).matmul(inverse_diagonal)
```

Comment:
No need to do the `diag_embed` here and to a matmul with a lot of zeros. You should be doing sth like the following (the unsqueezes might be wrong but you get the idea). 

```suggestion
            inverse_sqrt_diagonal = torch.diagonal(sample_as_covariance, dim1=-2, dim2=-1).clamp(min=1e-6).pow(-0.5)
            return inverse_diagonal.unsqueeze(-1) * sample_as_covariance * inverse_diagonal.unsqueeze(-2)
```